### PR TITLE
[LC-1933] Add challenge-based AI Passport authentication

### PR DESCRIPTION
**Pull request by TaylorBeeston**
## Summary

Replaces caller-supplied AI Passport DIDs with challenge-based DID Auth while keeping a deterministic compatibility path for the current backend. LearnCard signs a minimal VP, keeps the returned opaque bearer in module memory, sends it only through `Authorization`, and mints a fresh one-time ticket for every WebSocket connection.

Paired backend PR: https://github.com/WeLibraryOS/AI-Passport/pull/65

## Review findings resolved

- Adds `Content-Type: application/json` to session finalization.
- Clears stale Insights thread state before authentication and socket creation.
- Makes one reconnect state machine own retry accounting, one timer, cancellation, and the exact five-attempt budget.
- Replaces the durable bearer WebSocket protocol with `ai-passport-ticket.<one-time-ticket>`; the URL never contains either credential.
- Reads consent `challenge` and `domain` from React Router state and fails closed on malformed pairs.
- Rejects cross-origin URLs in authenticated fetches before credentials can be sent.
- Adds focused regression coverage for every item above.

## Compatibility and rollout

| Client | Backend | Outcome |
|---|---|---|
| This LearnCard PR | Current backend | Works through legacy transport only when `/auth/challenge` deterministically returns 404, 405, or the current backend's exact `401 { error: "Authentication required" }` response. |
| This LearnCard PR | Paired hardened backend | Works through challenged DID Auth, bearer REST, and one-time-ticket WebSockets. |
| Old LearnCard | Paired hardened backend | Breaks with `401`; the backend intentionally rejects legacy caller-supplied DID input. |

Deploy LearnCard first. Confirm the compatible release is live, complete the backend ownership migration, then deploy AI Passport. Network failures, malformed responses, and 5xx responses never trigger legacy fallback. LC-2115 tracks removal of the compatibility path after the rollback window.

## Verification

- `bun run test -- src/helpers/aiPassportAuth.test.ts src/stores/nanoStores/chatStore.test.ts` — 48 passed.
- `bunx vitest run src/pages/consentFlow/issueConsentFlowDidAuth.test.ts` — 5 passed.
- `bunx nx build learn-card-app` — passed.
- Full `learn-card-base` suite — 57 files passed; one unrelated date-boundary test failed because `calculateAge('2000-01-01', 2020-01-01T00:00:00-05:00)` expects 20 while the UTC date is still 2019-12-31.
- Paired Chromium flow with cookies cleared after exchange — bearer REST and a real WebSocket handshake succeeded; negotiated protocol was exactly `ai-passport`.

## Security notes

The bearer remains in module memory and within the normal application XSS boundary. WebSocket tickets are short-lived, origin-bound, session-bound, single-use, and freshly minted on reconnect. Authenticated fetches cannot target a different origin.

Fixes LC-1933.

### ✨ PR Description
Purpose: Implement challenge-based DID Auth for AI Passport authentication with backward-compatible fallback to legacy DID-in-query transport for pre-challenge backends.

Main changes:
- Added `aiPassportAuth` helper with session negotiation, challenge signing, and legacy detection supporting both bearer tokens and cookie-based authentication
- Replaced direct DID query parameters with `aiPassportFetch` wrapper that conditionally includes DID only for legacy backends during authentication negotiation
- Refactored consent flow redirects to use `getConsentFlowDidAuthRedirect` helper generating challenge-bound or delegated presentations with origin validation

_Generated by LinearB AI and added by gitStream._
<sub>AI-generated content may contain inaccuracies. Please verify before using.
💡 **Tip:** You can customize your AI Description using **Guidelines** [Learn how](https://docs.gitstream.cm/automation-actions/#describe-changes)</sub>

### CODE REVIEW

**Comment by TaylorBeeston** [comment]
Addressed the latest review findings in `0b6d0b6d2`:

- `finishSession` now sends JSON with the correct content type.
- Insights clears the previous thread before auth/socket setup.
- React Router location is the source for consent challenge/domain state.
- Reconnect attempts are counted once by one state machine; deliberate disconnect cancels the timer and stale async connects.
- Durable session bearers no longer ride in WebSocket protocols. Each connection mints a fresh one-time ticket from the paired backend.
- Authenticated fetches fail before `fetch` when a URL resolves off the configured AI Passport origin.

Focused tests cover the exact retry ceiling/reset/cancellation behavior, stale Insights frames, JSON headers, router query edge cases, ticket freshness, and credential non-exposure.

-- ChatGPT Work

**Comment by TaylorBeeston** [comment]
CI follow-up: the first E2E attempt failed before checkout/tests because the EC2 start endpoint returned `502` on all retries. I reran the failed job; the full E2E job passed in 25m6s. All automated LearnCard checks now pass. Chromatic still shows its manual “resume testing/review” gates.

-- ChatGPT Work

File: packages/learn-card-base/src/helpers/aiPassportAuth.ts [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+};
+
+export const aiPassportFetch = async (
+    path: string,
+    init: RequestInit = {},
+    did?: string
+): Promise<Response> => {
+    getAiPassportFetchUrl(path);
+
+    if (did && !getAiPassportAuthMode(did)) await authRequests.get(getAuthKey(did));
```
[the remark this answers is not in the document]
Comment (Custard7) [reply packages/learn-card-base/src/helpers/aiPassportAuth.ts]: Addressed in b238a4e2c — `aiPassportFetch` no longer proceeds unauthenticated when a caller skips the preflight. It now awaits any pending negotiation, otherwise establishes a session itself via the cached wallet (falling back to `walletStore`), and throws `'AI Passport authentication requires an initialized wallet'` only when no wallet exists to negotiate with. This self-heals the mutation callers (`usePreloadAssessment`, `useFinishAssessmentMutation`, `useUploadFileMutation`) that never called `ensureAiPassportSession` first, instead of turning them into hard failures. Covered by new unit tests.

**Comment by Custard7** [comment]
Addressed the remaining review findings in b238a4e2c, and verified the client contract against the paired backend PR (WeLibraryOS/AI-Passport#65):

**Fixed**
- **Stale durable bearers after logout (issue_8)**: new `clearAiPassportAuth()` wipes negotiated modes, bearer tokens, cached wallets, and in-flight negotiations. Wired into `handleAppLogout` alongside `disconnectWebSocket()` + `resetChatStores()`, so neither a session bearer nor an authenticated socket outlives the account that minted it.
- **Unauthenticated fetch hole (issue_10)**: `aiPassportFetch` with a DID but no negotiated mode now awaits the pending negotiation or establishes a session itself (cached wallet → `walletStore` fallback), throwing only when no wallet exists. Previously it silently sent the request with neither bearer nor legacy `?did=`.
- **Null-socket 5s stall (issue_9)**: `waitForSocketConnection` fails fast with `'WebSocket connection cancelled'` when a deliberate disconnect cancelled the in-flight connect (no socket, no pending connect, no scheduled reconnect), instead of blocking for the full timeout. All call sites already route rejections to the connection-error path.
- **Legacy-detection contract**: documented on `isLegacyChallengeResponse` that the hardened backend must never return `401 {error:"Authentication required"}` from `POST /auth/challenge`, since that exact shape triggers a silent downgrade to legacy transport.

**Already resolved (no change needed)**
- **Double reconnect increment (issue_7)**: `ws.onclose` only schedules; `reconnectWebSocket` is the sole owner of the counter. Verified single increment per attempt with a hard cap at `MAX_RECONNECT_ATTEMPTS`.

**Backend contract verification (AI-Passport#65)** — full match on all points:
1. Endpoints: `POST /auth/challenge`, `GET`/`POST /auth/session`, `POST /auth/websocket-ticket` all exist with matching methods.
2. Shapes: challenge `{audience, binding, challenge}`, JWT VP with `proofPurpose: authentication` + `domain: audience`, session `{authenticated, did, token}`, ticket `base64url(32 bytes)` (matches the client's `{32,128}` regex).
3. `/auth/challenge` only ever returns 200/403/405 — it has no 401 path, so false legacy detection cannot occur.
4. WS: one-time ticket travels only via `Sec-WebSocket-Protocol` (`ai-passport-ticket.<t>`), server echoes only `ai-passport`; the durable bearer is read exclusively from `Authorization`/cookie and never parsed from the subprotocol.
5. Error codes: protected routes emit 401 on missing/expired session, matching the client's single re-auth retry; challenge/session hard failures (400/403) correctly throw rather than downgrade.

New unit tests cover the logout clear, preflight-skipping callers, and the no-wallet rejection (56 tests passing across `aiPassportAuth`, `chatStore`, `credentialIngestion`).

**Review by Custard7 — APPROVED** [review]

**Comment by Custard7** [comment]
@TaylorBeeston this works well, I was able to verify locally. I pushed up some small fixes. ✅

File: packages/learn-card-base/src/helpers/aiPassportAuth.ts [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        }
+
+        const session = (await sessionResponse.json()) as SessionResponse;
+
+        if (!session.authenticated || session.did !== did || typeof session.token !== 'string') {
+            throw new Error('AI Passport session identity mismatch');
+        }
+
+        authModes.set(key, 'session');
+        authTokens.set(key, session.token);
```
[the remark this answers is not in the document]
Comment (TaylorBeeston) [reply packages/learn-card-base/src/helpers/aiPassportAuth.ts]: Fixed in `879eee5c7`. Session identity validation now checks only `authenticated` + the expected DID. The bearer is stored only when `token` is present, and any previously cached bearer for that auth key is removed first. Added a regression covering a valid cookie-only response and confirming the following authenticated request does not send `Authorization`.

-- ChatGPT Work

File: packages/learn-card-base/src/stores/nanoStores/chatStore.ts [hunk 3]
```diff
@@ -1534,9 +1669,10 @@ export async function closeInsightsSession(threadId?: string) {
...
     if (!activeThreadId) return;
 
-    const socket = connectWebSocket();
+    const socket = await connectWebSocket();
+
     if (!socket || socket.readyState !== WebSocket.OPEN) {
-        onReady(() => closeInsightsSession(activeThreadId));
+        onReady(() => void closeInsightsSession(activeThreadId));
         return;
     }
```
[the remark this answers is not in the document]
Comment (TaylorBeeston) [reply packages/learn-card-base/src/stores/nanoStores/chatStore.ts]: Fixed in `ac2edd342`. `closeInsightsSession` now sends only through the currently open socket. It no longer calls `connectWebSocket` or registers an `onReady` retry, so a deliberate disconnect cannot re-arm the reconnect state machine. The regression disconnects an open Insights socket, calls close, advances the reconnect window, and confirms no new socket or close frame appears.

-- ChatGPT Work

**Comment by TaylorBeeston** [comment]
Addressed the latest LearnCard review pass in `879eee5c7` and `ac2edd342`:

- Cookie-only session responses are accepted; optional bearers are retained only when present.
- Challenged consent callbacks must match the configured AI Passport origin before signing. Legacy unchallenged contract callbacks keep their existing behavior.
- Session starts disconnect before auth negotiation, eliminating the generation race.
- The paired backend can request fragment response mode, keeping the signed VP out of the callback request URL while old backends continue using the legacy query transport.
- `closeInsightsSession` no longer opens/reconnects a socket after deliberate disconnect.
- Every consent redirect path that previously allowed a signing/validation rejection to escape now shows a user-visible error toast. `FullScreenConsentFlow` already had this boundary.

I kept the exact legacy `401 { error: "Authentication required" }` capability fallback intentionally. It is limited to `POST /auth/challenge`; network, 5xx, malformed, and other 401 responses do not downgrade. The paired hardened backend has no 401 challenge path. Removing this case now would break the required LearnCard-first deployment against the current backend; LC-2115 tracks removing it after the rollback window.

Verification:
- focused LearnCard auth/chat/consent tests: 60 passed
- follow-up chat regression: 43 passed
- `nx build learn-card-app`: passed

-- ChatGPT Work

**Comment by TaylorBeeston** [comment]
Follow-up in `3c8763044` adds the missing stale-bearer transition regression. The test now:

1. authenticates through challenge/session and caches `stale-token`,
2. sends that bearer on a protected request that returns 401,
3. performs a fresh challenge/session exchange whose success is cookie-only,
4. retries the original protected request and proves `Authorization` is absent.

This covers the actual stale-bearer eviction path rather than only starting from an empty cache. The focused `aiPassportAuth` suite passes all 12 tests.

-- ChatGPT Work
